lib/charms/vault_k8s/v0/vault_kv.py

```diff
@@ -351,6 +371,53 @@ def get_credentials(self, relation: ops.Relation) -> dict:
         """Get the unit credentials from the relation."""
         return json.loads(relation.data[self.charm.app].get("credentials", "{}"))
 
+    def get_outstanding_kv_requests(self, relation_id: Optional[int] = None) -> List[KVRequest]:
+        """Get the outstanding requests for the relation."""
+        outstanding_requests: List[KVRequest] = []
+        kv_requests = self.get_kv_requests(relation_id=relation_id)
+        for request in kv_requests:
+            if not self._credentials_issued_for_request(nonce=request.nonce, relation_id=relation_id):
+                outstanding_requests.append(request)
+        return outstanding_requests
+
+    def get_kv_requests(self, relation_id: Optional[int] = None) -> List[KVRequest]:
+        """Get all KV requests for the relation."""
+        kv_requests: List[KVRequest] = []
+        relations = (
+            [
+                relation
+                for relation in self.model.relations[self.relation_name]
+                if relation.id == relation_id
+            ]
+            if relation_id is not None
+            else self.model.relations.get(self.relation_name, [])
+        )
+        for relation in relations:
+            app_data = dict(relation.data[relation.app])
+            for unit in relation.units:
+                unit_data = dict(relation.data[unit])
+                if not is_requirer_data_valid(app_data=app_data, unit_data=unit_data):
```

Comment:
I saw those `dict` creation from the relation data and noticed that the typing for `is_requirer_data_valid` specifies `dict`. We might be able to change those type hints to `Mapping` instead, and we could save creating new dictionaries from the relation data and just use the returned mapping directly. This is pretty minor however, so I will not block the PR for this.

Reply:
Done